File: 855-count-unique-characters-of-all-substrings-of-a-given-string/count-unique-characters-of-all-substrings-of-a-given-string.cpp
class Solution {
    #define mod 1000000007
public:
    int uniqueLetterString(string s) {
        int n = s.size();
        vector<vector<int>> hashmap(26);
        
        //Push -1 for all keys in value list
        for(int i=0;i<26;++i)   hashmap[i].push_back(-1);
        
        //Store all indices
        for(int i=0;i<n;++i)    hashmap[s[i]-'A'].push_back(i);
        
        //Push length of string for all keys at the end of each value list
        for(int i=0;i<26;++i)   hashmap[i].push_back(n);
        
        long long count = 0;
        for(int i=0;i<26;++i){
            for(int j=1;j<hashmap[i].size()-1;++j)
                count += (hashmap[i][j]-hashmap[i][j-1]) * (hashmap[i][j+1]-hashmap[i][j]);
        }
        return count;
    }
};

File: 855-count-unique-characters-of-all-substrings-of-a-given-string/count-unique-characters-of-all-substrings-of-a-given-string.java
public class Solution {
    final int MOD = 1000000007;

    public int uniqueLetterString(String s) {
        int n = s.length();
        List<List<Integer>> hashmap = new ArrayList<>(26);

        // Initialize lists with -1 for all keys
        for (int i = 0; i < 26; ++i) {
            List<Integer> list = new ArrayList<>();
            list.add(-1);
            hashmap.add(list);
        }

        // Store all indices
        for (int i = 0; i < n; ++i) {
            char c = s.charAt(i);
            hashmap.get(c - 'A').add(i);
        }

        // Push the length of the string for all keys at the end of each value list
        for (int i = 0; i < 26; ++i) {
            List<Integer> list = hashmap.get(i);
            list.add(n);
        }

        long count = 0;
        for (int i = 0; i < 26; ++i) {
            List<Integer> list = hashmap.get(i);
            for (int j = 1; j < list.size() - 1; ++j) {
                count += (list.get(j) - list.get(j - 1)) * (list.get(j + 1) - list.get(j));
            }
        }

        return (int) (count % MOD);
    }
}